File: OC版本/CDDSlideTabBarDemo/CDDSlideTabBarDemo/DCSlideTabBarController.m
//
//  DCSlideTabBarController.m
//  CDDSlideTabBarDemo
//
//  Created by SnailChen on 2018/10/17.
//Copyright © 2018年 SnailChen. All rights reserved.
//

/******************    TabBar          *************/
#define MallClassKey   @"rootVCClassString"
#define MallTitleKey   @"title"
#define MallImgKey     @"imageName"
#define MallSelImgKey  @"selectedImageName"

#import "DCSlideTabBarController.h"
// Controllers

// Models
#import "DCSlideBarViewAnimation.h"
// Views

// Vendors

// Categories

// Others

@interface DCSlideTabBarController ()<UITabBarControllerDelegate>



@end

@implementation DCSlideTabBarController

#pragma mark - LazyLoad


#pragma mark - LifeCyle
- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    
    //遵守代理
    self.delegate = self;
    
    [self addChildViewContorller];
}


#pragma mark - 添加子控制器
- (void)addChildViewContorller
{
    NSArray *childArray = @[
                            @{MallClassKey  : @"DCFViewController",
                              MallTitleKey  : @"美信",
                              MallImgKey    : @"tabr_01_up",
                              MallSelImgKey : @"tabr_01_down"},
                            
                            @{MallClassKey  : @"DCSViewController",
                              MallTitleKey  : @"首页",
                              MallImgKey    : @"tabr_02_up",
                              MallSelImgKey : @"tabr_02_down"},
                            
                            @{MallClassKey  : @"DCTViewController",
                              MallTitleKey  : @"美媒榜",
                              MallImgKey    : @"tabr_03_up",
                              MallSelImgKey : @"tabr_03_down"},
                            ];
    [childArray enumerateObjectsUsingBlock:^(NSDictionary *dict, NSUInteger idx, BOOL * _Nonnull stop) {
        
        UIViewController *vc = [NSClassFromString(dict[MallClassKey]) new];
        UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:vc];
        UITabBarItem *item = nav.tabBarItem;
        item.image = [UIImage imageNamed:dict[MallImgKey]];
        item.selectedImage = [[UIImage imageNamed:dict[MallSelImgKey]] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
        item.imageInsets = UIEdgeInsetsMake(6, 0,-6, 0);//（当只有图片的时候）需要自动调整
        [self addChildViewController:nav];
    }];
}



#pragma mark - 点击动画
- (void)tabBarButtonClick:(UIControl *)tabBarButton
{
    for (UIView *imageView in tabBarButton.subviews) {
        if ([imageView isKindOfClass:NSClassFromString(@"UITabBarSwappableImageView")]) {
            //需要实现的帧动画,这里根据自己需求改动
            CAKeyframeAnimation *animation = [CAKeyframeAnimation animation];
            animation.keyPath = @"transform.scale";
            animation.values = @[@1.0,@1.1,@0.9,@1.0];
            animation.duration = 0.3;
            animation.calculationMode = kCAAnimationCubic;
            //添加动画
            [imageView.layer addAnimation:animation forKey:nil];
        }
    }
    
}

- (UIControl *)getTabBarButton{
    NSMutableArray *tabBarButtons = [[NSMutableArray alloc]initWithCapacity:0];
    
    for (UIView *tabBarButton in self.tabBar.subviews) {
        if ([tabBarButton isKindOfClass:NSClassFromString(@"UITabBarButton")]){
            [tabBarButtons addObject:tabBarButton];
        }
    }
    UIControl *tabBarButton = [tabBarButtons objectAtIndex:self.selectedIndex];
    return tabBarButton;
}


#pragma mark - <UITabBarControllerDelegate>

- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController{
    
    [self tabBarButtonClick:[self getTabBarButton]]; //点击tabBarItem动画代理
}

- (id<UIViewControllerAnimatedTransitioning>)tabBarController:(UITabBarController *)tabBarController animationControllerForTransitionFromViewController:(UIViewController *)fromVC toViewController:(UIViewController *)toVC{
    
    //页面切换代理
    NSArray *tabViewControllers = tabBarController.viewControllers;
    
    NSLog(@"toVc : %zd  ----- fromVc : %zd",[tabViewControllers indexOfObject:toVC],[tabViewControllers indexOfObject:fromVC]);
    
    if ([tabViewControllers indexOfObject:toVC] > [tabViewControllers indexOfObject:fromVC]) { // left
        return [[DCSlideBarViewAnimation alloc] dcInitWithSlideTargetEdge:UIRectEdgeLeft];
    }else{ // right
        return [[DCSlideBarViewAnimation alloc] dcInitWithSlideTargetEdge:UIRectEdgeRight];
    }
}

@end

File: Swift版本/CDDSlideTabBarDemo/CDDSlideTabBarDemo/DCSlideTabBarController.swift
//
//  DCSlideTabBarController.swift
//  CDDSlideTabBarDemo
//
//  Created by 陈甸甸 on 2019/2/23.
//Copyright © 2019 陈甸甸. All rights reserved.
//


import UIKit

class DCSlideTabBarController: UITabBarController{
    
    
    // MARK: - LazyLoad
    var currentChildVc: UIViewController?
    var areaInsets: CGFloat = 0.0
    
    
    // MARK: - LifeCycle
    override func viewDidLoad() {
        super.viewDidLoad()

        setUpUI()
        
        addChildViewContorller() //设置子控制器
    }
}

// MARK: - 设置 UI 界面
extension DCSlideTabBarController {
    
    fileprivate func setUpUI() {
        
        view.backgroundColor = .white
        
        self.delegate = self
        
        /// 安全距离的简单判断
        areaInsets = (UIScreen.main.bounds.size.height >= 812) ? 34.0 : 0
    }
}



// MARK: - 设置子控制器
extension DCSlideTabBarController {
    
    func addChildViewContorller () {
        
        let childArray = [
            [
                "rootVCClassString"  : "DCFViewController",
                "title"  : "美信",
                "imageName"    : "tabr_01_up",
                "selectedImageName" : "tabr_01_down"
            ],[
                "rootVCClassString"  : "DCSViewController",
                "title"  : "美信",
                "imageName"    : "tabr_01_up",
                "selectedImageName" : "tabr_01_down"
            ],[
                "rootVCClassString"  : "DCTViewController",
                "title"  : "美信",
                "imageName"    : "tabr_01_up",
                "selectedImageName" : "tabr_01_down"
            ]
            ] as [Any]
        
        for tabDict in childArray {
            
            let result = tabDict as![String:Any]  // 先声明告诉tabDict是个字典
            
            guard let vcName = result["rootVCClassString"] as? String else { continue }
            guard let normalImg = result["imageName"] as? String else { continue }
            guard let selImg = result["selectedImageName"] as? String else { continue }
//            guard let title = result["title"] as? String else { continue }
            
            let childVc = getVcFromString(vcName)
            currentChildVc = childVc
            
            childVc.tabBarItem.image = UIImage(named: normalImg)
            childVc.tabBarItem.selectedImage = UIImage(named: selImg)
            childVc.tabBarItem.imageInsets = UIEdgeInsets(top: 6, left: 0, bottom: -6, right: 0) //只有图片的时候需要设置 大致偏移6px
//            childVc.tabBarItem.title = title //tu
            
            let childNav = UINavigationController(rootViewController: childVc)
            
            addChild(childNav)

        }
        
        if let currentChildVc = currentChildVc { // 处理tabBarItem 选中颜色
            guard let tabBar = currentChildVc.tabBarController?.tabBar else {
                return
            }
            guard let count = currentChildVc.tabBarController?.children.count else {
                return
            }
            let tabBarWidth = tabBar.frame.size.width
            let tabBarHeight = tabBar.frame.size.height
            let childCount = CGFloat(count)
            
            tabBar.selectionIndicatorImage = drawImageWithColor(color: UIColor.darkGray, size: CGSize(width: tabBarWidth / childCount, height: tabBarHeight + areaInsets))
        }
    }
    
    //  swiftlint:disable force_unwrapping
    private func getVcFromString(_ vcName: String) -> UIViewController {
        
        guard let nameSpace = Bundle.main.infoDictionary!["CFBundleExecutable"] as? String else {
            print("没有获取到命名空间")
            return UIViewController()
        }
        
        guard let childVcClass = NSClassFromString(nameSpace + "." + vcName) else {
            print("没有获取到字符串对应的Class")
            return UIViewController()
        }
        
        guard let childVcType = childVcClass as? UIViewController.Type else {
            print("没有获取对应控制器的类型")
            return UIViewController()
        }
        
        return childVcType.init()
    }
    
    
    /// 绘制选中Item的背景色
    private func drawImageWithColor(color : UIColor , size : CGSize) -> UIImage {
        guard  size.width > 0 && size.height > 0  else {
            return UIImage()
        }
        let itemRect = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        UIGraphicsBeginImageContextWithOptions(itemRect.size, false, 0) //开始绘制
        let contRef : CGContext = UIGraphicsGetCurrentContext()!;
        
        contRef.setFillColor(color.cgColor);
        contRef.fill(itemRect);
        let selectBgImage = UIGraphicsGetImageFromCurrentImageContext() ?? UIImage(); //绘图
        UIGraphicsEndImageContext(); //结束绘制
        
        return selectBgImage
    }
}

// MARK: - 处理代理方法
extension DCSlideTabBarController : UITabBarControllerDelegate {

    /// 点击动画
    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
        
        tabBarButtonClick(tabBarButton: getTabBarButton())
        
    }
    
    
    /// 滑动动画
    func tabBarController(_ tabBarController: UITabBarController, animationControllerForTransitionFrom fromVC: UIViewController, to toVC: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        
        return DCSlideBarAnimation() //滑动动画
    }
    

    private func getTabBarButton() -> UIControl {

        var tabBarButtons = Array<UIView>()
        for tabBarButton in tabBar.subviews {
            if (tabBarButton.isKind(of:NSClassFromString("UITabBarButton")!)){
                tabBarButtons.append(tabBarButton)
            }
        }
        return tabBarButtons[selectedIndex] as! UIControl
    }
    
    private func tabBarButtonClick(tabBarButton : UIControl) {
        
        for imageView in tabBarButton.subviews {
            if (imageView.isKind(of: NSClassFromString("UITabBarSwappableImageView")!)){
                let animation = CAKeyframeAnimation()
                animation.keyPath = "transform.scale"
                animation.duration = 0.3
                animation.calculationMode = CAAnimationCalculationMode(rawValue: "cubicPaced")
                animation.values = [1.0,1.1,0.9,1.0]
                imageView.layer.add(animation, forKey: nil)
            }
        }
    }
    
}

